### PR TITLE
ASoC: SOF: force end-of-file for debugfs trace at suspend

### PR DESCRIPTION
**Pull request by kv2019i**
Current trace implementation gets out of sync when sof device is
put to suspend. The debugfs file handle is kept open, but
firmware will reset its state. After resume, debugfs client's read
offset will not be synchronized to firmware and this may result
in traces read in incorrect order and/or stale data being read
after resume.

Changelog:
- V5 7th May: new bug introduced in V4, still more work to improve patch and description understandability
- V4 26th Apr: Split into multiple patches, dropped renames.
- V3 24th Apr: Updated patchset with simplified code. Reworked to signal EOF only once so that this works with old and new sof-loggers. 
- V2 11th Apr: Revised patchset to address comments so far, no longer RFC
- V1 3th Aprh: initial patchset (RFC)

Related patches:
- sof-logger patch PR1312 (need not be merged at the same time)
   https://github.com/thesofproject/sof/pull/1312

~~ Issues handled in previous iterations~~:
 - ~~this is not full alternative but has overlap with issue #306 (and related PR)~~ -> this can be merged as incremental update
 - ~~locking needs to be revisited still (I'll go through this if we move forward)~~ -> handled in PR800
 - ~~having multiple u32 entries for various single-bit pieces of dtrace info in sof-priv.h is ugly, I'd like to combine these~~ -> sorted in 11th April push

### CODE REVIEW

File: sound/soc/sof/trace.c [hunk 1]
```diff
@@ -30,6 +30,12 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
 	if (sdev->host_offset > pos)
 		return sdev->host_offset - pos;
 
+	/* No data available and trace disabled -> eof */
+	if (!sdev->dtrace_is_enabled && sdev->host_offset > 0) {
```
Comment (ranj063) [inline sound/soc/sof/trace.c]: @kv2019i is the second check necessary? sdev->host_offset > 0?

File: sound/soc/sof/trace.c [hunk 2]
```diff
@@ -81,6 +89,12 @@ static ssize_t sof_dfsentry_trace_read(struct file *file, char __user *buffer,
 		return -EIO;
 	}
 
+	/* end-of-file reached, force position to zero */
+	if (avail == 0) {
```
Comment (ranj063) [inline sound/soc/sof/trace.c]: @kv2019i Is this whole if block needed, if avail is 0, count below will be 0 and we will copy nothing and return 0 anyway isnt it?

File: sound/soc/sof/trace.c [hunk 3]
```diff
@@ -95,8 +109,21 @@ static ssize_t sof_dfsentry_trace_read(struct file *file, char __user *buffer,
...
 
+static int sof_dfsentry_trace_open(struct inode *inode, struct file *file)
+{
+	struct snd_sof_dfsentry *dfse = inode->i_private;
+	struct snd_sof_dev *sdev = dfse->sdev;
+
+	if (inode->i_private)
+		file->private_data = inode->i_private;
+	sdev->dtrace_eof = false;
+	sdev->host_offset = 0;
```
Comment (ranj063) [inline sound/soc/sof/trace.c]: @kv2019i should the host_offset be reset too?

**Comment by ranj063** [comment]
@singalsu can you please give this a try and see if it fixes the problem you faced with missing traces with runtime PM enabled?

File: sound/soc/sof/trace.c (hunk 1 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @ranj063 The second check is still needed as the trace client will reopen the debugfs node and we don't want to repeatedly signal eof (we do if the client repeatedly opens the file and seeks to some non-zero position and tries to read). 

Without this check, 'sof-logger -t'  will end up in a loop where it continuously opens the debugfs node, tries to read, gets eof, closes, and loop restarts.

With the check in place, sof-logger is forced to reopen the device, and its first read at position 0 will block until FW has data available. And this ensure that when the FW writes to position 0, user-space and kernel are in sync.

File: sound/soc/sof/trace.c (hunk 2 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @ranj063 No, you are right. The code would work by just setting the EOF flag here and continue code.  I ended up putting return directly here as it seemed more readable (as we have direct returns elsewhere in the functin as well).

Comment (ranj063) [reply sound/soc/sof/trace.c]: @kv2019i why will we need to set eof here? would we reach here only if the trace_sleep is woken up and eof is already set in the sof_wait_trace_avail()?

File: sound/soc/sof/trace.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @ranj063 Uhm, yes, but I do do reset "sdev->host_offset" here, right? :) Or did you mean the reverse that it should _not_ be reset here?

If you refer to the file node position, that's something we don't really have much control over, which makes this a lot more complicated in general. The sof-logger implementation does a lot of seeks, so we need to keep supporting arbitrary seeks (although sof-logger's seeks are sequential, but it still does seeks). So our debugfs read() just has to work with whatever position the client wants to read from.

My patch now covers this for most part. One major downside is that traces disappear when runtime resume is done. So if sof-logger doesn't read the traces in time, then you lose those. But with typical timeouts for runtime-pm, sof-logger has enough time to get the traces read out. 

And of course, this does not help to get late traces between suspend-resume. But you'll anyways have a blackout period when the trace is not set up so FW cannot push the traces anywhere.

Comment (ranj063) [reply sound/soc/sof/trace.c]: @kv2019i sorry I meant the hsot_offset should not be reset here. we reset the host_offset when we resume from suspend. Could this possibly why you're losing the trace from before the suspend?

Comment (kv2019i) [reply sound/soc/sof/trace.c]: @ranj063 I tested without, but we hit one corner case if we don't reset at open:

- FW is in runtime suspend, trace inactive
- trace client opens but sdev->host_offset is set to N (N > 0, from previous run)
- trace read waiting for data
- FW is resumed, tracing starts at offset 0
- trace client is woken up and it will start reading data from [N....buffer-size] -> you get stale traces read out

We can avoid this by forcing reset of sdev->host when the trace client is opened.

**Comment by singalsu** [comment]
@ranj063 I just tried this with the sof patch rebuilt tools but now I get empty trace output. I'll check tomorrow with Kai what I've missed.

File: sound/soc/sof/trace.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/trace.c]: @kv2019i keyon's PR is exactly about this case. We need to start copying the new trace upon resuming from N instead of 0.

**Comment by singalsu** [comment]
@ranj063 @kv2019i Correction, I didn't realize that despite the patched trace didn't show the prior content from boot until this moment when start it shows it later!

Now when I had left sof-logger on and started music playback it traced out normally and I saw now the params() output that got dropped previously when started. Also the traces from boot are seen (component new() stuff) is seen when I scroll back the screen.

So, this looks usable to me now and improvement from non-patched version. Though the start of showing the events should be tested with some lower amount of output test case. Maybe pressing volume up/down from alsamixer and see if it activates the trace.

**Comment by ranj063** [comment]
@singalsu thanks for confirming. I think the missing piece is @keyonjie 's work to copy trace starting from host_offset instead of 0 upon resume.

File: sound/soc/sof/trace.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @ranj063 wrote: 
> @kv2019i keyon's PR is exactly about this case. We need to start copying the new trace upon resuming from N instead of 0.

Correct. This patch is not full replacement to Keyon's PR.

But this is worth considering. E.g. currently when you do S3 suspend, sof-logger gets disconnected. Trace client can reconnect by reopening the trace file, but it has no way to know where it should seek to when it opens. So you'll in practise get 0...64kB traces as duplicate when S3 suspend cycle occurs. This problem remains with Keyon's patch as well. With my patch, sof-logger -t output is consistent both over runtime-suspend and s3 suspend (but you have the limitations discussed above).

With Keyon's patch, you need to solve the s3 suspend issue plus implement the offset logic for BYT fw, before merging the whole thing.

OTOH, one downside is that you cannot easily use "cat" to capture the trace data during a use-case as the cat will terminate at each runtime suspend. With Keyon's patch + some fix to S3, you could avoid this.

Btw, for even more complete support, something like kernel's scheduler trace might be a better model (top/kernel/trace/trace.c and top/fs/seq_file.c) -- with seeking allowed in our interface, covering all the cases can be quite complicated but seq_file.c can help to reuse existing logic for that.

Anyways, we need to take sensible incremental steps to improve the system. This RFC is one possible such increment, or we agree to jump to offset support directly.

File: sound/soc/sof/trace.c (hunk 2 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @ranj063 wrote:
> @kv2019i why will we need to set eof here? would we reach here only if the trace_sleep is woken up and eof is already set in the sof_wait_trace_avail()?

There are two cases how we hit eof:
1) trace client sleeping in read -> runtime suspend -> eof for trace with above code
2) trace client not sleeping -> runtime suspend -> eof signalled when all pending data is read by trace app 

Hmm, I can think whether this could be done in one place, but for now I had to have the eof marking in two places to cover all cases.

File: sound/soc/sof/trace.c (hunk 3 above)
Comment (ranj063) [reply sound/soc/sof/trace.c]: @kv2019i agree with a step-by-step approach. But offset support should be added soon enough.

File: sound/soc/sof/sof-priv.h [hunk 4]
```diff
@@ -407,6 +407,8 @@ struct snd_sof_dev {
 	int dma_trace_pages;
 	wait_queue_head_t trace_sleep;
 	u32 host_offset;
+	/* FIXME: combine dtrace flags into one enum */
+	bool dtrace_eof;
```
Comment (keyonjie) [inline sound/soc/sof/sof-priv.h]: we have too many DMA trace related in this root snd_sof_dev structure, and we actually need more comments to describe how they work, worthy to refine them especially when we plan to provide disabling DMA trace option, but we can do it next.

File: sound/soc/sof/trace.c (hunk 3 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c]: @kv2019i here reset sdev->host_offset will lead to issues like this:
each time you open "sof-logger -t", the logs is empty.
That means only logs after tool run are guaranteed will be captured and output. For those logs at the first boot(user never have change to run sof-logger before that) e.g. topology parsing, pipeline creation, they themselves don't have chance to be showed to you.

Please be noticed that even end user never run "sof-logger -t", the sdev->host_offset is keep updating, so when you run the tool and reset it, you will break this updating.

File: sound/soc/sof/trace.c (hunk 1 above)
Comment (keyonjie) [reply sound/soc/sof/trace.c]: when we go to this point, it means no more logs to be read at the moment(sdev->host_offset == pos), so if dma trace is disabled, we can set sdev->dtrace_eof directly. OTOH, "sdev->host_offset > 0" means nothing actually, it is only an offset.

Comment (kv2019i) [reply sound/soc/sof/trace.c]: @keyonjie Maybe this needs a easier to understand implementation, but in the current patch, "sdev->host_offset > 0" does have a meaning. Basicly we know "sdev->host_offset" is reset to zero at open, so this check allows us to detect the case that dtrace 1) has been been active (as it is non-zero, 2) has since been stopped (!dtrace_is_enabled), and 3) all data has been read by client (host_offset==pos). With 1+2+3 met, we can mark as eof and force the client to reopen. After it reopens, host_offset will be reset to zero and this check will no longer bet hit.

But yup, as both you and Ranjani had questions about this, I think the patch needs work to handle this case.

UPDATE: reworked this part, closing this conversation.

File: sound/soc/sof/trace.c (hunk 3 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @keyonjie That is true, this is probably biggest side-effect of this RFC patch. You can workaround this by delaying module load at boot, start sof-logger and only then modprobe SOF module, but true, this does change the behaviour compared to current state.

If we don't do this, we are stuck with the problem that at open(), we have no way to synchronize the positions between client and FW and we hit more issues:
- with suspend/resume cycles, you will get repeatedly old traces in the output,
- if the FW has rolled over, you will get [0...host_offset] traces (and not the full 64KB history FW has written to the buffer

... so not running sof-logger while FW is active, is anyways problematic and without extending the buffering scheme, this is only reliable if less than 64K of traces is written before sof-logger is started.

I'll think about if we can make some special case to handle the early traces without breaking the other logic.

UPDATE: solved in a later patchset, early logs are now available

File: sound/soc/sof/sof-priv.h (hunk 4 above)
Comment (kv2019i) [reply sound/soc/sof/sof-priv.h]: Ack, agreed this would be good to do.

**Comment by kv2019i** [comment]
@keyonjie wrote:
> before suspend, I have concern about reset sdev->host_offset at file open, we need figure out that.

Ok, I figured out a way to capture the early logs if sof-logger is started after FW has been active. Please see my comments above I wrote today, and the new patch I uploaded today.

**Comment by kv2019i** [comment]
>     * 24th Apr: Updated patchset with simplified code. Reworked to signal EOF only once so that this works with old and new sof-loggers.

New patchset available. This tries to simplify the logic while still keeping the main feature of resyncing position at suspend/resume cycles. A new sof-logger patch was needed (see PR1312).

File: sound/soc/sof/sof-priv.h [hunk 5]
```diff
@@ -411,10 +411,15 @@ struct snd_sof_dev {
 	struct snd_dma_buffer dmatb;
 	struct snd_dma_buffer dmatp;
 	int dma_trace_pages;
-	wait_queue_head_t trace_sleep;
 	u32 host_offset;
-	u32 dtrace_is_enabled;
-	u32 dtrace_error;
+	unsigned int dtrace_is_enabled:1; /* FW DMA trace is active */
+	unsigned int dtrace_error:1;      /* FW sent non-recoverable error */
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h]: do we care about optimizing two bitfields?

File: sound/soc/sof/sof-priv.h [hunk 6]
```diff
@@ -411,10 +411,15 @@ struct snd_sof_dev {
...
+
+	/* DFS trace client state */
+	wait_queue_head_t dftrace_sleep;
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h]: Not sure I see the  point of renaming this? we have a mix of 'trace', 'dtrace', 'dftrace' and it starts to look messy?

File: sound/soc/sof/trace.c [hunk 7]
```diff
@@ -31,23 +30,37 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
...
+	/* data immediately available */
+	if (ret)
+		return ret;
+
+	/* tracing has ended, signal EOF to client */
+	if (!sdev->dtrace_is_enabled && sdev->dftrace_eof) {
+		sdev->dftrace_eof = 0;
+		return 0;
+	}
+
```
Comment (plbossart) [inline sound/soc/sof/trace.c:54]: I would first split this function in two in a separate patch, then add the new functionality and add the renames last if needed. It's really confusing to review as is.

File: sound/soc/sof/trace.c [hunk 8]
```diff
@@ -97,10 +110,22 @@ static ssize_t sof_dfsentry_trace_read(struct file *file, char __user *buffer,
...
+		sdev->host_offset = 0;
+
+	return 0;
+}
+
 static const struct file_operations sof_dfs_trace_fops = {
 	.open = simple_open,
 	.read = sof_dfsentry_trace_read,
 	.llseek = default_llseek,
+	.release = sof_dfsentry_trace_release,
```
Comment (plbossart) [inline sound/soc/sof/trace.c:132]: this looks odd since we don't have a matching function where host_offset is initialized to zero? I would also add this as a separate patch to separate out the changes.

File: sound/soc/sof/sof-priv.h (hunk 5 above)
Comment (kv2019i) [reply sound/soc/sof/sof-priv.h]: @plbossart This was more to set usage for future fields -- does not seem reasonable to keep adding u32 fields. But true, I was originally planning to add more fields, but now that I'm only adding one, it does indeed seem out of place. I'll drop.

File: sound/soc/sof/sof-priv.h (hunk 6 above)
Comment (kv2019i) [reply sound/soc/sof/sof-priv.h]: @plbossart Hmm, I wanted to split state we have towards FW ("DMA trace" -> dtrace) and state related to the debugfs trace node ("debugfs trace" -> dftrace). Now the two are a bit mixed in the code, making it harder to follow. But I take it did not help. I'll drop.

File: sound/soc/sof/trace.c (hunk 8 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c:132]: @plbossart Here the debugfs versus fwtrace state confusion is apparent. 'host_offset' is controlled by the DMA trace state machine.

If debugfs node is opened while DMA trace is active, we don't really have good solutions how to sync the positions (debufgs client will get data from 0....host_offset).

If debugfs node is opened while DMA trace is not active, we don't want to reset 'host_offset' as the user may want to read whatever traces are in the log (@keyonjie commented on this earlier).

But by resetting 'host_offset' in release when DMA trace is not active, we can at least ensure the same data is passed to client twice.

So long story short, this is a bit iffy, but is there for a reason. I'll add a comment to this, let's see how it reads out.

File: sound/soc/sof/trace.c [hunk 9]
```diff
@@ -13,10 +13,9 @@
 #include "sof-priv.h"
```
Comment (plbossart) [inline sound/soc/sof/trace.c:13]: Commit message:

ASoC: SOF: trace: remove code duplication in sof_wait_trace_avail()

Move duplicated code in sof_wait_trace_avail() to a helper function

File: sound/soc/sof/sof-priv.h [hunk 10]
```diff
@@ -415,6 +415,8 @@ struct snd_sof_dev {
 	u32 host_offset;
 	u32 dtrace_is_enabled;
 	u32 dtrace_error;
+	u32 dtrace_eof:1;		/* trace EOF pending */
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h]: the bitfield is useless here, you'll still allocate 32 bits :-)

File: sound/soc/sof/trace.c [hunk 11]
```diff
@@ -43,6 +43,12 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
 	if (ret)
 		return ret;
 
+	/* tracing has ended, signal EOF to client */
+	if (!sdev->dtrace_is_enabled && sdev->dtrace_eof) {
+		sdev->dtrace_eof = 0;
```
Comment (plbossart) [inline sound/soc/sof/trace.c]: 'Signal EOF' by settings dtrace_eof to zero? That sounds counter-intuitive. Usually state_xyz = 0 means you are not in that xyz state.

Comment (plbossart) [inline sound/soc/sof/trace.c]: I am missing what this new dtrace_eof actually does. It looks like a state value but I don't see any action taken nor any 'notification to client'. What am I missing?

Comment (kv2019i) [reply sound/soc/sof/trace.c]: @plbossart EOF is signalled by returning 0 available bytes to read and waking up user-space. 'dtrace_eof' is internal bookkeeping so that we only signal once (otherwise an older sof-logger binary would go to a 100% CPU busy-loop after this patch is merged).

Comment (kv2019i) [reply sound/soc/sof/trace.c]: @plbossart See above, dtrace_eof is internal booking, a request to signal eof, the actual EOF signal is returning 0 from read(). I moved the comment within the braces, maybe that clarifies the intent better.

File: sound/soc/sof/trace.c (hunk 9 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c:13]: @plbossart Ack, fixed.

File: sound/soc/sof/sof-priv.h (hunk 10 above)
Comment (kv2019i) [reply sound/soc/sof/sof-priv.h]: @plbossart That is true. Ok, I'll modify. :)

File: sound/soc/sof/sof-priv.h [hunk 12]
```diff
@@ -415,6 +415,8 @@ struct snd_sof_dev {
 	u32 host_offset;
 	u32 dtrace_is_enabled;
 	u32 dtrace_error;
+	u32 dtrace_eof;		/* trace EOF pending */
```
Comment (plbossart) [inline sound/soc/sof/sof-priv.h]: dtrace_eof is set at the init (twice), so I am confused if we really have a 'trace EOF pending' during regular execution. Something's not quite self-explanatory here.

File: sound/soc/sof/trace.c [hunk 13]
```diff
@@ -43,6 +43,12 @@ static size_t sof_wait_trace_avail(struct snd_sof_dev *sdev,
 	if (ret)
 		return ret;
 
+	if (!sdev->dtrace_is_enabled && sdev->dtrace_eof) {
+		sdev->dtrace_eof = false;
```
Comment (plbossart) [inline sound/soc/sof/trace.c]: once this dtrace_eof is false, when does it become true again, if ever?

File: sound/soc/sof/trace.c [hunk 14]
```diff
@@ -182,6 +204,7 @@ int snd_sof_init_trace_ipc(struct snd_sof_dev *sdev)
 	}
 
 	sdev->dtrace_is_enabled = true;
+	sdev->dtrace_eof = true;
```
Comment (plbossart) [inline sound/soc/sof/trace.c]: is it necessary to initialize dtrace_eof here, as well as .....

File: sound/soc/sof/trace.c [hunk 15]
```diff
@@ -291,6 +314,8 @@ void snd_sof_release_trace(struct snd_sof_dev *sdev)
 			"error: fail in snd_sof_dma_trace_release %d\n", ret);
 
 	sdev->dtrace_is_enabled = false;
+	sdev->dtrace_eof = true;
```
Comment (plbossart) [inline sound/soc/sof/trace.c]: .... here. dtrace_is_enabled is force to false at the beginning of the function, then set here. No such assignment for dtrace_eof so not sure if this init is needed.

**Comment by kv2019i** [comment]
@plbossart wrote:
> Sorry @kv2019i still confused about the expected transitions for the dtrace_eof boolean. this might be a missing set of comments to explain how this might be used.

Eh, sorry, this round was waste of your time, I slipped a bug in the V4 patchset that further obfuscated intended usage of dtrace_eof variable (although functionally still ok and passed my tests). No excuses, should have been more careful.

Even without my mistake, I agree 'dtrace_eof' is still not ideal, too many comments are needed to explain it. I went back and forth with different names and ended up to rename the variable to 'dtrace_draining' in V5 patchset and further iterated the comments (shorter, just focusing on WHY).

This is still a kludge, I agree, but has a lot of practical utility for sof-logger usage, so I'll still update this PR. If you think this has no chance in upstream, then we might consider dropping altogether and wait until we have time to implement a bigger rework (current WA is to disable runtime PM when taking traces).

File: sound/soc/sof/trace.c (hunk 13 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @plbossart The eof (draining in latest) is set true whenever FW tracing is disabled. This happens e.g. every time runtime PM kicks in and suspends the DSP. We (might) have traces in DDR that have not bee read by user-space. We want to give user-space a chance to read these traces out, but not get confused when DSP resumes and starts writing from 0 again.

File: sound/soc/sof/trace.c (hunk 14 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @plbossart So this is the stupid mistake I made. Sorry about that again. That should have been "sdev->dtrace_eof = false" -- when FW trace is active, we have no need to signal EOF of course.

File: sound/soc/sof/trace.c (hunk 15 above)
Comment (kv2019i) [reply sound/soc/sof/trace.c]: @plbossart This code is intentional. At end of function, FW tracing is disabled so we mark the eof/draining bit -> we want to give user-space a chance to read the data out before resume kicks in (-> trace training in latest patchset).

**Comment by ranj063** [comment]
@kv2019i something is not right with this methodology suggested in this PR. This is what I tried.
1. After boot: ./sof-logger -l sof-apl.ldc -t
2. stop logger and start again -> no output
3. resume DSP with aplay 
4. run sof-logger and now I can see the output again. 
5. stop logger and start it again - > no output 

So everytime I stop the logger and start it again its output is empty. This is likely because you reset the host_offset to 0 in trace_release. I find this behaviour odd though. I think sof-logger should always be able to read the trace irrespective of whether it is opened for the first time after the DSP is suspended or closed & opened again.

**Comment by keyonjie** [comment]
Acked on @ranj063, please make sure the PR really works as we expected before we can merge it, the debugfs trace is used to share FW logs for debugging, so the criteria here should be if it really helps for the debugging, we need show the logs as much as possible, resetting the host_offset and clearing the log output at each logger open sounds really bad idea to me.

**Comment by kv2019i** [comment]
Ranjani wrote:
>     1. After boot: ./sof-logger -l sof-apl.ldc -t
>     2. stop logger and start again -> no output 
>     3. resume DSP with aplay
>     4. run sof-logger and now I can see the output again. 
>     5. stop logger and start it again - > no output
[...]
> So everytime I stop the logger and start it again its output is empty. This is likely because 
> you reset the host_offset to 0 in trace_release. I find this behaviour odd though. I think 
> sof-logger should always be able to read the trace irrespective of whether it is opened for 
> the first time after the DSP is suspended or closed & opened again.

@ranj063 and @keyonjie , yes, this is exactly the expected behaviour. I sent an RFC about the change of behaviour to the SOF mailing list in April when I started to work on this:
https://mailman.alsa-project.org/pipermail/sound-open-firmware/2019-April/003277.html

Note you will not lose traces. You'll always get the last (upto 64K) traces FW sent even if you did not have logger running.

So with the patch, the DMA trace works like "dmesg -c" in kernel. You can get logs out reliably, but you can only read them out once. Until we have implemented offset trace (or trace double-buffering), we must reset the host_offset so SW/FW are resycnhronized. Currently you will often get duplicated traces, or not traces at all -> it's very undeterministic now.

**Comment by kv2019i** [comment]
@keyonjie wrote:
> the debugging, we need show the logs as much as possible, resetting the host_offset and clearing the 
> log output at each logger open sounds really bad idea to me.

Ack, but this is **not** done in the patch. We clear offset at close, specifically to allow user-space to read out the traces before we sync on the position.

**Comment by ranj063** [comment]
> So with the patch, the DMA trace works like "dmesg -c" in kernel. You can get logs out reliably, but you can only read them out once. Until we have implemented offset trace (or trace double-buffering), we must reset the host_offset so SW/FW are resycnhronized. Currently you will often get duplicated traces, or not traces at all -> it's very undeterministic now.

@kv2019i but this is quite inconvenient if you can only read the trace once after the DSP is suspended. We can live with it for a short while but we need the bigger change to implement trace offset and should be done soon. Users will start complaining about not being able to read the trace the second time around.

**Comment by kv2019i** [comment]
@ranj063 wrote:
> @kv2019i but this is quite inconvenient if you can only read the trace once after the DSP is suspended. > We can live with it for a short while but we need the bigger change to implement trace offset and 
> should be done soon. Users will start complaining about not being able to read the trace the second 
> time around.

But in context of this PR, the question is which is _now_ a bigger problem:
  1) getting duplicate or no traces in normal case unless you disable runtime PM, or
  2) inability to to read the trace node multiple times

That's the call we need to make for this PR -> @ranj063 @keyonjie @plbossart 
I'm arguing (1) is the bigger problem and am requesting merging this.

Longer term, I'll reiterate my stance that the offset patch does not solve all issues. The trace buffer remains a scratch buffer and even with offset support implemented, user-space has no way to discover where the DSP is writing to. So only reliable way to use DMA trace is to run sof-logger all the time and store traces to user-space continuously. Offset support will not fix this.

**Comment by ranj063** [comment]
> I'm arguing (1) is the bigger problem and am requesting merging this.

@kv2019i I personally feel we can't deal with either of them. A few months ago I'd have agreed with you and said (1) is the bigger problem but now we've got external folks using SOF and it is not easy to explain to them that the logger can only read once with PM enabled.

**Comment by ranj063** [comment]
@kv2019i I have a simple suggestion. I agree this PR is necessary but its annoying to not be able to read the data the second time around. Can we possibly change the logger application to read and save the data at first read (possibly in a file) and read from there instead?

**Comment by kv2019i** [comment]
@ranj063 wrote:
> you and said (1) is the bigger problem but now we've got external folks using SOF and it is not easy to > explain to them that the logger can only read once with PM enabled.

But why the expectation of a "trace store"?  Can you elaborate the usage model a bit. For debugging, you should be storing traces continuously to a file. If you want to debug an error after something strange happens, you should cp out the files (trace, exception, memory dumps) and analyze offlien (and attach to bug issue). What is the usage where you read the trace may times? 

I think the interface should be viewed as a "trace port" with some intermediate buffering, but it's not a trace store and you'll end with issues if you expect it to be a store (might work sometimes, but not reliable).

**Comment by kv2019i** [comment]
@ranj063 wrote:
> read the data the second time around. Can we possibly change the logger application to read and 
> save the data at first read (possibly in a file) and read from there instead?
 
But this is already supported, right. You 'cp /sys/kernel/debug/sof/trace logfile' and run sof-logger with '-i logfile'. Having sof-logger cache data implicitly sounds a complicated. Where to write, how many traces to keep, potential confusion of combining invalid data from state file cache and valid data from a target data, etc, etc. If really, really need, I'd add a carbon-copy option where sof-logger copies the verbatim binary data all the data to a file, in addition to showing out the parsed log.

**Comment by ranj063** [comment]
> @ranj063 wrote:
> 
> > read the data the second time around. Can we possibly change the logger application to read and
> > save the data at first read (possibly in a file) and read from there instead?
> 
> But this is already supported, right. You 'cp /sys/kernel/debug/sof/trace logfile' and run sof-logger with '-i logfile'. Having sof-logger cache data implicitly sounds a complicated. Where to write, how many traces to keep, potential confusion of combining invalid data from state file cache and valid data from a target data, etc, etc. If really, really need, I'd add a carbon-copy option where sof-logger copies the verbatim binary data all the data to a file, in addition to showing out the parsed log.

@kv2019i sorry didnt realize we had the option of reading from a file with the logger. I am OK with it as long we shout out loud and clear that this is a handicap with the logger at the moment. 
@keyonjie what do you think?

**Comment by kv2019i** [comment]
> @keyonjie what do you think?

@keyonjie ping, see above

**Comment by keyonjie** [comment]
Sorry @kv2019i didn't have much time to go to details of your commits and explaining, I am seeing DMA trace logs lost issue(sometimes no any DMA trace logs show even the FW is running and theoretically we should get logs e.g. trigger commands, hw_params commands) when doing KWD debugging these days.

Just went through your commits, my understand to it is something like this:
1) snd_sof_init_trace() // reset host_offset, set dtrace_is_enabled, clear dtrace_draining.
2) file->open() // nothing special we do for trace.
3) file->read() 
        sof_wait_trace_avail() // if has enough data, return; otherwise, wait for event.
4) file->release() // e.g. Ctrl+c to exit sof-logger. 
5) snd_sof_release_trace() // clear dtrace_is_enabled, set dtrace_draining.

Without suspend/resume, the typical flow is 1->2->3->4(first sof-logger run)->2->3->4...->2->3->4 (the Nth sof-logger run), this works fine as dtrace_draining is never set and we do nothing in 4) actually.  dtrace_is_enabled is true, dtrace_draining is false, and host_offset is keep aligned with FW position update.

When suspend happens, let's consider about 2 scenarios:

a. suspend without sof-logger run, that is 1->[2->3->4]->5, after 5), dtrace_is_enabled is false, dtrace_draining is true, and host_offset is aligned with the last update from FW position update.

Here, when we run sof-logger, 2->3(read out all logs, and then clear dtrace_draining, works fine here)->4(reset host_offset, that means if we try open sof-logger again, no logs will be showed at the 2nd iteration, this is also fine IMO).

Then 1->[2->3->4] again, everything same with the first boot, we won't get logs that belong to the previous active stage, this also looks legit(so for first boot fail issues e.g. tplg IPC failure, we can run sof-logger before we turn the driver back to resume via any audio commands, to debug those issues).

**So in general, this case a is fine for me.**

b. suspend with sof-logger run, that is 1->2->3->5, after 5), dtrace_is_enabled is false, dtrace_draining is true, and host_offset is aligned with the last update from FW position update.

Here, the last logs before suspend is already showed, so if 
        b_1. 4) is run(Ctrl+c) at suspended, we will reset host_offset(No problem, we have already read them).
        b_2. go back to 1) with resume(sof-logger keep running), the next trace position update from FW will start from 0(might be >0 actually), then the sof-logger will output staled(or not existed) [host_offset, buffer_size) logs before it is aligned to the new logs data, **this is a little bit confused here, but more logs/noise is better than no logs, I can be tolerant to this actually**.

**So this case b is acceptable to me.**

So, after long analyzing and chatter here, I think this is good PR to be merged now, thank you so much.

**Comment by kv2019i** [comment]
Thank you @keyonjie for the very detailed response and review. Your analysis is correct. I only have one note for this: 
> b. suspend with sof-logger run, that is 1->2->3->5, after 5), dtrace_is_enabled is false, dtrace_draining is true, and host_offset is aligned with the last update from FW position update.
[...]
> b_2. go back to 1) with resume(sof-logger keep running), the next trace position update from FW will start from 0(might be >0 actually), then the sof-logger will output staled(or not existed) [host_offset, buffer_size) logs before it is aligned to the new logs data, **this is a little bit confused here, but more logs/noise is better than no logs, I can be tolerant to this actually**.

In fact we can cover this case as well. With recent patches to sof-logger, when you run sof-logger in trace mode (-t) and this b_2 case happens, sof-logger will notice the end of draining phase (read returns once with 0 bytes available) and reopens the trace node. So when FW starts again, sof-logger will continue reading from 0.

If you use an older sof-logger binary, then the result will be as per your analysis above and some stale traces will be output.

> So, after long analyzing and chatter here, I think this is good PR to be merged now, thank you so much.

Great, thank you! @plbossart This would seem to be good to go!

**Comment by keyonjie** [comment]
> > b_2. go back to 1) with resume(sof-logger keep running), the next trace position update from FW will start from 0(might be >0 actually), then the sof-logger will output staled(or not existed) [host_offset, buffer_size) logs before it is aligned to the new logs data, **this is a little bit confused here, but more logs/noise is better than no logs, I can be tolerant to this actually**.
> 
> In fact we can cover this case as well. With recent patches to sof-logger, when you run sof-logger in trace mode (-t) and this b_2 case happens, sof-logger will notice the end of draining phase (read returns once with 0 bytes available) and reopens the trace node. So when FW starts again, sof-logger will 

Great to know about this, thanks.

> continue reading from 0.
>

**Comment by plbossart** [comment]
Thanks @keyonjie let's merge and test
